napari/utils/events/evented_model.py

```diff
@@ -55,15 +71,118 @@ def no_class_attributes():
     def _return2(x, y):
         return y
 
-    main.ClassAttribute = _return2
+    pydantic_main.ClassAttribute = _return2
     try:
         yield
     finally:
         # undo our monkey patch
-        main.ClassAttribute = utils.ClassAttribute
+        pydantic_main.ClassAttribute = pydantic_utils.ClassAttribute
 
 
-class EventedMetaclass(main.ModelMetaclass):
+class ParametrizedGenericCompliantModelField(ModelField):
+    """
+    A ModelField that correcly interprets and validates a wider variety
+    of parametrized generics.
+
+    See:
+    - https://github.com/napari/napari/pull/4609
+    - https://github.com/samuelcolvin/pydantic/issues/4161
+    """
+
+    def _type_analysis(self):
+        super()._type_analysis()
+        origin = get_origin(self.outer_type_)
+
+        # if class validators are unset, it means pydantic ignored them, so
+        # we add them back in just like pydantic does
+        if not self.class_validators:
+            get_validators = getattr(origin, '__get_validators__', dict)
+            self.class_validators.update(
+                {
+                    f'{origin.__name__}_{i}': Validator(validator, pre=True)
+                    for i, validator in enumerate(get_validators())
+                }
+            )
+
+        # since issubclass(collections.abc.Set, typing.Set) == False, and our EventedSet
+        # is a collections.abc.Set, we have to tell pydantic to treat it the same as the other sets
+        # isinstance(origin, type) is needed for non-class origins (e.g: Literal)
+        if (
+            origin is not None
+            and isinstance(origin, type)
+            and issubclass(origin, abc.Set)
+        ):
+            self.shape = SHAPE_SET
+            self.type_ = get_args(self.outer_type_)[0]
+
+    def populate_validators(self):
+        super().populate_validators()
+        # we need to get rid of the basic arbitrary_types validator because otherwise
+        # in cases where we don't coerce type (e.g: Colormap) it always fails cause the type is wrong
+        # We then coerce the type ourselves below
+        self.validators = [
+            f
+            for f in self.validators
+            if f.__name__ != 'arbitrary_type_validator'
+        ]
+
+        origin = get_origin(self.outer_type_) or self.outer_type_
+        # anything that's not really a class should not be used for coercion
+        if (
+            not isinstance(origin, type)
+            or getattr(origin, '_is_protocol', False)
+            or not getattr(origin, '_coerce_type', True)
```

Comment:
I don't love this construct.  I just pulled and played with it a bit.  I started by commenting it out in all the settings fields ... but no tests failed (so that's one note... it would be easy for someone to remove them without knowing)

then I removed on `Array` to see the fail... the stack trace was extremely hard to parse... had I not been specifically reviewing this PR, I'm not sure I would have had any idea that the "solution" would have been to add back `_coerce_type` or indeed that it even had anything to do with logic added in this class at all.

I think the addition of _coerce type in `conint` also has a code smell. (hard to figure out why this would have any bearing on a very minor extension of a builtin pydantic object).

this just generally feels like something that "got this PR to work" ... but doesn't have the feel of a framework that napari developers will readily understand and employ correctly.  

It also feels like it overlaps a bit (or specifically opts out of) with the intention of pydantic in the first place.  I'm sure there's a good reason, but it's not immediately clear what this opting out here couldn't happen in the validator method of the type (rather than our own additional API on top of pydantic).  would love to get some more background/explanation on this and consider alternatives

Reply:
> I started by commenting it out in all the settings fields ... but no tests failed

Some tests should definitely fail, that's why I needed to add it. Specifically, the `yaml` dumping of the settings failed because `Language` is not coerced to a normal `str`, and the dumper does not know what to do with it.

> then I removed on Array to see the fail... the stack trace was extremely hard to parse... 

Yes, this is just a general gripe of pydantic; the validation errors are super hard to chase down.

> I'm not sure I would have had any idea that the "solution" would have been to add back _coerce_type or indeed that it even had anything to do with logic added in this class at all.
>
> I think the addition of _coerce type in conint also has a code smell. (hard to figure out why this would have any bearing on a very minor extension of a builtin pydantic object).

I generally agree, and the solution would be to switch from an opt-out to an opt-in, but the consequence is that we lose coercion anything that is, for example, a `Sequence` that we didn't specifically subclass for this purpose. But maybe it's not that much work, and I can try to implement it that way to see what breaks down.

> It also feels like it overlaps a bit (or specifically opts out of) with the intention of pydantic in the first place.

I disagree with this: pydantic is simply inconsistent with type coercion (subclasses of `str` just become `str`, some things return different types depending on their parametrization, other things are simply never coerced). This PR is essentially unifying this behaviour (or trying to), but there are unintended consequences because we are using *types* as a way to re-use validators across models, rather than simply having validators on each model.

> it's not immediately clear what this opting out here couldn't happen in the validator method of the type (rather than our own additional API on top of pydantic).

Because pydantic (sometimes, inconsistently) adds its own validators at the end of user-made ones, which override whatever type coercion we did.

Reply:
> Some tests should definitely fail, that's why I needed to add it. Specifically, the yaml dumping of the settings failed because Language is not coerced to a normal str, and the dumper does not know what to do with it.

🤷‍♂️ ... they didn't fail.  perhaps there's something being skipped on macos, or required CI, or something

Reply:
> Is this additional coerce type thing mandatory for the original goal of this PR (to fix parametrized generics)? Could it be considered on its own? Or are they inextricably linked?

They are linked: the problem with parametrized generics is that they fall in that category of things that get coerced to a "base" type (so no matter what you do in you `Sequence` subclass, it will always end up being a `list` if you parametrize it, and your validators will be ignored).

> man_shrugging ... they didn't fail. perhaps there's something being skipped on macos, or required CI, or something

:D bizarre! Anyways, it's not *always strictly necessary* to add `_coerce_type = False`. It only is if that value is then passed around to places where having, for example, a `class Language(str)` causes issues.

Reply:
I just pushed a change to swap the logic: it's now opt-in. Notice that while I had to add the coercion to some things (Evented objects) it's not actually necessary everywhere: `ColorArray` is fine without it because it has coercion logic in its validator AND it's one of those cases where `class_validators` are respected.

Reply:
thanks, I'm having a play with `populate_validators` myself right now.  need to understand it a bit better before I can say anything intelligent :joy: